## 1. Import  libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

## 2. Import a new customer data set

In [3]:
# Import data

path = r'/Users/berk/Instacart_Grocery_Basket_Analysis'

In [4]:
df_customers = pd.read_csv(os.path.join(path, '02_Data', 'Original_Data', 'customers.csv'))

In [5]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

## 3. Data Wrangling

### Missing values count

In [6]:
# Check the customers' dataset for any missing data.

df_customers.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

There are 11259 rows where the first name is missing

### Renaming

In [7]:
# Changing the names of the columns to make it easier to understand the variable

df_customers.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'last_name', 'Gender' : 'gender', 
                        'STATE' : 'state', 'Age' : 'age', 'n_dependants' : 'number_dependents'}, inplace = True)

In [8]:
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_dependents fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374

### Changing data types

In [9]:
# Checking the column type 'user id'

df_customers['user_id'].dtype

dtype('int64')

In [10]:
# Changing the data type of a variable

df_customers['user_id'] = df_customers['user_id'].astype('str')

In [11]:
df_customers['user_id'].dtype

dtype('O')

## 4. Data Consistency Checks

In [12]:
# Data Consistency Checks df_customers dataframe

df_customers.describe()

age  number_dependents         income
count  206209.000000      206209.000000  206209.000000
mean       49.501646           1.499823   94632.852548
std        18.480962           1.118433   42473.786988
min        18.000000           0.000000   25903.000000
25%        33.000000           0.000000   59874.000000
50%        49.000000           1.000000   93547.000000
75%        66.000000           3.000000  124244.000000
max        81.000000           3.000000  593901.000000

### Check for mixed-type data 

In [13]:
# Check for mixed-type data in your df_customers dataframe

for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
        print (col, ' mixed')
  else: print(col, ' consistent')

user_id  consistent
first_name  mixed
last_name  consistent
gender  consistent
state  consistent
age  consistent
date_joined  consistent
number_dependents  consistent
fam_status  consistent
income  consistent


first_name column has mixed-type data 

In [14]:
# Changing column type to str

df_customers['first_name'] = df_customers['first_name'].astype('str')

In [15]:
df_customers['first_name'].dtype

dtype('O')

### Check for duplicate values

In [16]:
# Run a check for duplicate values in your df_customers data

df_customers_dups = df_customers[df_customers.duplicated()]

In [17]:
df_customers_dups.size

0

The customer dataset has no duplicate values.

## 5. Merging datasets

### Import the orders_products_merged_updated_2

In [18]:
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02_Data','Prepared_Data', 'orders_products_merged_updated_2.pkl'))

In [19]:
df_ords_prods_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ...    price_range_loc     busiest_day    busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regularly days   
1          1  ...  Mid-range product  Regularly busy    Slowest days   
2          1  ...  Mid-range product  Regularly busy    Slowest days   
3          1  ...  Mid-range product      Least busy    Slowest days   
4          1  ...  Mid-range product      Least busy    Slowest days   

   busiest_period_of_day  max_order  loyalty_flag avg_price spending_flag  \
0         Average orders         10  New customer      14.0  High spender   
1         Average orders         10  New customer      14.0  High spender   
2            Most orders         10  New customer      14.0  High spender   
3         Average orders         10  New customer      14.0  High spender   
4            Most orders         10  New customer      14.0  High spender   

  median_prior_orders   order_frequency_flag  
0                20.5  Non-frequent customer  
1                20.5  Non-frequent customer  
2                20.5  Non-frequent customer  
3                20.5  Non-frequent customer  
4                20.5  Non-frequent customer  

[5 rows x 26 columns]

In [20]:
df_ords_prods_merged = df_ords_prods_merged.drop(columns = ['eval_set'])

In [21]:
df_ords_prods_merged.shape

(32404859, 25)

In [22]:
# Changing column type to str


df_ords_prods_merged['user_id'] = df_ords_prods_merged['user_id'].astype('str')

In [23]:
df_ords_prods_merged['user_id'].dtype

dtype('O')

In [24]:
df_ords_prods_custs = df_ords_prods_merged.merge(df_customers, on = 'user_id')

In [25]:
df_ords_prods_custs.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0  ...   order_frequency_flag  first_name  last_name  gender  \
0         195  ...  Non-frequent customer       Linda     Nguyen  Female   
1         195  ...  Non-frequent customer       Linda     Nguyen  Female   
2         195  ...  Non-frequent customer       Linda     Nguyen  Female   
3         195  ...  Non-frequent customer       Linda     Nguyen  Female   
4         195  ...  Non-frequent customer       Linda     Nguyen  Female   

     state age date_joined number_dependents fam_status  income  
0  Alabama  31   2/17/2019                 3    married   40423  
1  Alabama  31   2/17/2019                 3    married   40423  
2  Alabama  31   2/17/2019                 3    married   40423  
3  Alabama  31   2/17/2019                 3    married   40423  
4  Alabama  31   2/17/2019                 3    married   40423  

[5 rows x 34 columns]

In [26]:
# checking if the merge function worked successfully

df_ords_prods_custs.shape

(32404859, 34)

## 6. Export this new dataframe

In [27]:
df_ords_prods_custs.to_pickle(os.path.join(path, '02_Data','Prepared_Data', 'df_ords_prods_custs.pkl'))